### PR TITLE
[RSDK-7310] Update the board interface

### PR DESCRIPTION
**Pull request by penguinland**
This turned out to be two problems combined! 1) update the SDK to use a more recent version of the protobufs in the API repo, and then 2) remove the `Status` RPC from the boards. The PR isn't quite done: I'm unsure how this should interact with https://github.com/viamrobotics/viam-cpp-sdk/pull/230, which is where the compiled version of the updated protos is. but the rest I think is ready for a look.

Updating the protobufs was harder than I expected: https://github.com/viamrobotics/api/pull/470 (unrelated to boards) added a new type that had not been used previously, which meant that getting it to work required updating CMakeLists.txt to include the Google libraries for the new type.

Changes to boards:
- `get_status` doesn't exist any more. You can get the standard status from a component the same way you would for a motor, but there isn't a special board-specific one any more.
- In the component status you _can_ get, analog and digital interrupt values are stored as ints, not wrappers around ints. No need to unwrap.
- `get_analog_names` and `get_digital_interrupt_names` can't easily be implemented any more. They don't exist in the Flutter and TypeScript SDKs, and are hopefully being removed from the Go and Python SDKs soon (they have a purpose on the server side, but no obvious use on the client side, so probably shouldn't be in these SDKs). So, I removed them. 
- Similarly, I removed `get_analog_readers` and `get_digital_interrupts`, as they're not present in the Go SDK and would be very hard to implement without the old status.

Everything compiles against a recent version of the API repo, but not against the protobufs copied into this repo itself. How should that be accomplished?

### CODE REVIEW

File: src/viam/examples/modules/complex/proto/buf.lock [hunk 1]
```diff
@@ -4,5 +4,4 @@ deps:
   - remote: buf.build
     owner: googleapis
     repository: googleapis
-    commit: ee48893a270147348e3edc6c1a03de0e
-    digest: shake256:a35b0576a2b55dad72747e786af05c03539c2b96be236c9de39fe10d551931ac252eb68445c0cef6bbd27fa20e8c26eee5b8a9fe9c2fde6f63a03e18f8cf980d
+    commit: 4ed3bc159a8b4ac68fe253218760d035
```
Comment (penguinland) [inline src/viam/examples/modules/complex/proto/buf.lock:7]: This file updated automatically at some point. I don't know what caused it, but I've put the change in a separate commit so it's easy to revert if it's not supposed to be here.

**Review by martha-johnston — APPROVED** [review]
as always, thanks for leaving such a detailed description, it was super helpful

File: src/viam/api/CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,7 @@ if (VIAMCPPSDK_USE_DYNAMIC_PROTOS)
 endif()
 
 set(BUF_PROTO_COMPONENTS
+  app/mltraining
```
Comment (acmorrow) [inline src/viam/api/CMakeLists.txt:22]: Agree that this is now needed due to https://github.com/viamrobotics/api/pull/470

File: src/viam/api/CMakeLists.txt [hunk 3]
```diff
@@ -178,12 +181,98 @@ if (VIAMCPPSDK_USE_DYNAMIC_PROTOS)
     ${PROTO_GEN_DIR}/component/servo/v1/servo.grpc.pb.h
     ${PROTO_GEN_DIR}/component/servo/v1/servo.pb.cc
     ${PROTO_GEN_DIR}/component/servo/v1/servo.pb.h
+    ${PROTO_GEN_DIR}/google/api/annotations.grpc.pb.cc
```
Comment (acmorrow) [inline src/viam/api/CMakeLists.txt:184]: The existing inclusion of things from `google/api` appears only to have pulled in the `.pb` files and not the `grpc.pb` files. Can you see if things still work if you filter out the `grpc` ones?

File: src/viam/api/CMakeLists.txt [hunk 4]
```diff
@@ -178,12 +181,98 @@ if (VIAMCPPSDK_USE_DYNAMIC_PROTOS)
...
+    ${PROTO_GEN_DIR}/google/api/expr/v1beta1/value.grpc.pb.h
+    ${PROTO_GEN_DIR}/google/api/expr/v1beta1/value.pb.cc
+    ${PROTO_GEN_DIR}/google/api/expr/v1beta1/value.pb.h
+    ${PROTO_GEN_DIR}/google/api/field_behavior.grpc.pb.cc
+    ${PROTO_GEN_DIR}/google/api/field_behavior.grpc.pb.h
+    ${PROTO_GEN_DIR}/google/api/field_behavior.pb.cc
+    ${PROTO_GEN_DIR}/google/api/field_behavior.pb.h
+    ${PROTO_GEN_DIR}/google/api/field_info.grpc.pb.cc
+    ${PROTO_GEN_DIR}/google/api/field_info.grpc.pb.h
+    ${PROTO_GEN_DIR}/google/api/field_info.pb.cc
```
Comment (acmorrow) [inline src/viam/api/CMakeLists.txt:238]: I know it is a pain to do, but this is a *lot* of new code and I'm sure a lot of it is entirely irrelevant to our usage. Could you try pruning each subtree under google/api and see if there are some swaths we can kill off?

**Review by acmorrow — CHANGES_REQUESTED** [review]
Reviewing only build stuff here, I'll leave the actual SDK change to other reviewers, though since they are almost entirely removals it basically looks fine to me.

**Review by stuqdog — COMMENTED** [review]
Generally looks good to me! It seems we're pulling in a lot of new code though, is it possible to filter some of that down?

File: src/viam/api/CMakeLists.txt [hunk 5]
```diff
@@ -353,6 +487,8 @@ set_target_properties(
   VERSION ${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}
 )
 
+target_link_options(viamapi PRIVATE "LINKER:-z,defs")
```
[the remark this answers is not in the document]
Comment (penguinland) [reply src/viam/api/CMakeLists.txt:490]: I have no clue about CMake. It's here because Drew used it to help debug linker problems, and should be removed now that we're done debugging. Sorry I can't explain more. :shrug:

File: src/viam/api/CMakeLists.txt [hunk 6]
```diff
@@ -178,12 +181,98 @@ if (VIAMCPPSDK_USE_DYNAMIC_PROTOS)
...
+    ${PROTO_GEN_DIR}/google/rpc/context/attribute_context.pb.cc
+    ${PROTO_GEN_DIR}/google/rpc/context/attribute_context.pb.h
+    ${PROTO_GEN_DIR}/google/rpc/error_details.grpc.pb.cc
+    ${PROTO_GEN_DIR}/google/rpc/error_details.grpc.pb.h
+    ${PROTO_GEN_DIR}/google/rpc/error_details.pb.cc
+    ${PROTO_GEN_DIR}/google/rpc/error_details.pb.h
+    ${PROTO_GEN_DIR}/google/rpc/status.grpc.pb.cc
+    ${PROTO_GEN_DIR}/google/rpc/status.grpc.pb.h
+    ${PROTO_GEN_DIR}/google/rpc/status.pb.cc
+    ${PROTO_GEN_DIR}/google/rpc/status.pb.h
```
[the remark this answers is not in the document]
Comment (penguinland) [reply src/viam/api/CMakeLists.txt:275]: This might be my mistake: I had assumed that all the generated grpc code needed to go in here, and this was all new since the last time the C++ SDK was synced with the API repo. I'll try pruning it down...

File: src/viam/examples/modules/complex/proto/buf.lock [hunk 7]
```diff
@@ -4,5 +4,4 @@ deps:
   - remote: buf.build
     owner: googleapis
     repository: googleapis
-    commit: ee48893a270147348e3edc6c1a03de0e
-    digest: shake256:a35b0576a2b55dad72747e786af05c03539c2b96be236c9de39fe10d551931ac252eb68445c0cef6bbd27fa20e8c26eee5b8a9fe9c2fde6f63a03e18f8cf980d
+    commit: 74015a8aeb8445aa9e3e1454cb54bc35
```
Comment (penguinland) [inline src/viam/examples/modules/complex/proto/buf.lock:7]: I still don't know what this file does, but it changed again. Any ideas?

**Review by penguinland — COMMENTED** [review]
Take another look: the extra build dependencies are pruned down to just:
- app/mltraining/v1/ml_training.pb.* (dependency added in https://github.com/viamrobotics/api/pull/470), and
- google/rpc/status.pb.*mentioned in ml_training.pb.*

I'm still unsure how this PR should interact with https://github.com/viamrobotics/viam-cpp-sdk/pull/230, to get the latest .proto files over here. Any advice?

File: src/viam/api/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (penguinland) [reply src/viam/api/CMakeLists.txt]: You were totally right, thanks for pushing back on this!

File: src/viam/api/CMakeLists.txt (hunk 3 above)
Comment (penguinland) [reply src/viam/api/CMakeLists.txt]: You're right: a lot of the generated protobuf files were unused and could be removed again!

File: src/viam/api/CMakeLists.txt [hunk 8]
```diff
@@ -287,6 +293,7 @@ target_sources(viamapi
     ${PROTO_GEN_DIR}/google/api/annotations.pb.cc
     ${PROTO_GEN_DIR}/google/api/http.pb.cc
     ${PROTO_GEN_DIR}/google/api/httpbody.pb.cc
+    ${PROTO_GEN_DIR}/google/rpc/status.pb.cc
```
Comment (acmorrow) [inline src/viam/api/CMakeLists.txt:296]: ⬇️ Very likely you also need to add `status.pb.h` down below line 340.

Comment (penguinland) [reply src/viam/api/CMakeLists.txt:296]: I agree that it looks like it should be required. I was able to build without that line, but it feels more proper to add it anyway. I've added it. Things continue to build when using a local copy of the protos, and I still don't know what the plan is to update the ones in this repo itself.

To rebuild things, I've been using this: `cd ..; rm -rf build; mkdir build; cd build; cmake -DVIAMCPPSDK_OFFLINE_PROTO_GENERATION=ON -DVIAMCPPSDK_USE_DYNAMIC_PROTOS=ON -DVIAMCPPSDK_USE_LOCAL_PROTOS=ON .. -G Ninja; ninja -j 5 all` and I imagine that the final version will use fewer options on the `cmake` part.

Comment (acmorrow) [reply src/viam/api/CMakeLists.txt:296]: You would only notice it to be missing if you tried to build a program against an installed build of the SDK.

**Comment by penguinland** [comment]
Github's CI cannot build this because it requires a more recent version of the compiled .proto files from the api repo than what's currently in `src/viam/api/`. What's the procedure for updating that? I still believe the correct versions might be in #230.

**Comment by penguinland** [comment]
I talked to Ethan about how to get the updated protos: he says I should rebase this branch off of #230, then change this PR to pull into that branch instead of `main`. When we're happy here, we merge this into #230, and then merge that PR into `main`. Changes will be coming soon...

**Comment by penguinland** [comment]
Okay, I've rebased and changed the target branch. Let's see how this looks...

**Review by acmorrow — APPROVED** [review]

**Comment by penguinland** [comment]
Rebased to get rid of merge conflicts. If the CI checks pass, I'm going to merge this...

File: src/viam/examples/modules/complex/proto/buf.lock (hunk 7 above)
Comment (stuqdog) [reply src/viam/examples/modules/complex/proto/buf.lock:7]: It sets the buf version for the example module but is setup to auto-update. it's safe to ignore!
